timeboost/s3_storage.go

```diff
@@ -0,0 +1,218 @@
+			return 5 * time.Second
+		}
+		s.lastFailedDeleteRound = 0
+	}
+
+	bids, round, err := s.sqlDB.GetBids(s.config.MaxDbRows)
+	if err != nil {
+		log.Error("Error fetching validated bids from sql DB", "round", round, "err", err)
+		return 5 * time.Second
+	}
+	// Nothing to persist or a contiguous set of bids wasn't found, so exit early
+	if len(bids) == 0 {
+		return s.config.UploadInterval
+	}
+
+	var csvBuffer bytes.Buffer
+	var size int
+	var firstBidId int
+	csvWriter := csv.NewWriter(&csvBuffer)
+	uploadAndDeleteBids := func(firstRound, deletRound uint64) error {
+		// End current batch when size exceeds MaxBatchSize and the current round ends
+		csvWriter.Flush()
+		if err := csvWriter.Error(); err != nil {
+			log.Error("Error flushing csv writer", "err", err)
+			return err
+		}
+		if err := s.uploadBatch(ctx, csvBuffer.Bytes(), firstRound); err != nil {
+			log.Error("Error uploading batch to s3", "firstRound", firstRound, "err", err)
+			return err
+		}
+		// After successful upload we should go ahead and delete the uploaded bids from DB to prevent duplicate uploads
+		// If the delete fails, we track the deleteRound until a future delete succeeds.
+		if err := s.sqlDB.DeleteBids(deletRound); err != nil {
+			log.Error("error deleting s3-persisted bids from sql db", "round", deletRound, "err", err)
+			s.lastFailedDeleteRound = deletRound
+		} else {
+			// Previously failed deletes dont matter anymore as the recent one (larger round number) succeeded
+			s.lastFailedDeleteRound = 0
+		}
+		return nil
+	}
+
+	header := []string{"ChainID", "Bidder", "ExpressLaneController", "AuctionContractAddress", "Round", "Amount", "Signature"}
+	if err := csvWriter.Write(header); err != nil {
+		log.Error("Error writing to csv writer", "err", err)
+		return 5 * time.Second
+	}
+	for index, bid := range bids {
+		record := []string{bid.ChainId, bid.Bidder, bid.ExpressLaneController, bid.AuctionContractAddress, fmt.Sprintf("%d", bid.Round), bid.Amount, bid.Signature}
+		if err := csvWriter.Write(record); err != nil {
+			log.Error("Error writing to csv writer", "err", err)
+			return 5 * time.Second
+		}
+		if s.config.MaxBatchSize != 0 {
+			size += csvRecordSize(record)
+			if size >= s.config.MaxBatchSize && index < len(bids)-1 && bid.Round != bids[index+1].Round {
+				if uploadAndDeleteBids(bids[firstBidId].Round, bids[index+1].Round) != nil {
+					return 5 * time.Second
+				}
+				// Reset csv for next batch
+				csvBuffer.Reset()
+				if err := csvWriter.Write(header); err != nil {
+					log.Error("Error writing to csv writer", "err", err)
+					return 5 * time.Second
+				}
+				size = 0
+				firstBidId = index + 1
+			}
+		}
+	}
+	if s.config.MaxBatchSize == 0 || size > 0 {
+		if uploadAndDeleteBids(bids[firstBidId].Round, round) != nil {
+			return 5 * time.Second
+		}
+	}
+
+	if s.lastFailedDeleteRound != 0 {
+		return 5 * time.Second
+	}
+
+	return s.config.UploadInterval
```

Comment:
Using sleep (essentially what CallIteratively is doing) for the UploadInterval will cause clock drift since each period will become UploadInterval + processing time. I think we should use a different approach to maintain consistent upload timing.

Reply:
I agree, but I think it is necessary that the uploads happen sequentially/iteratively even if we have to give up on accuracy due to processing.

Reply:
I think the current impl can be improved though, I'll use time.Ticker instead as it will allow us to continue uploading instantly if the previous processing took more time then the upload interval